File: libptiff/src/io/backend/memory_backend.cpp
#include <array>
#include <cstddef>
#include <cstdint>
#include <utility>
#include <vector>

#include <ptiff/io/backend/memory/memory_codec.hpp>
#include <ptiff/io/backend/memory/memory_image_sink.hpp>
#include <ptiff/io/backend/memory/memory_image_source.hpp>
#include <ptiff/io/backend/memory/memory_layout.hpp>
#include <ptiff/io/backend/memory_backend.hpp>
#include <ptiff/io/backend_factory.hpp>
#include <ptiff/io/binary_writer.hpp>

namespace ptiff::io::backend {

namespace {

// In-memory ("PMEM") binary header. Layout of byte 0 on:
//   [0..4)  magic "P","M","E","M" (0x50 0x4D 0x45 0x4D)
//   [4..6)  version, little-endian u16 == 1
//   [6..8)  flags, little-endian u16 == 0 (reserved)
//   [8..12) model-header byte length M (little-endian u32)
//   [12..)  model header (see the codec); then the contiguous per-image pixel regions.
constexpr std::array<std::byte, 4> kMagic{
    std::byte{0x50}, std::byte{0x4D}, std::byte{0x45}, std::byte{0x4D}};
constexpr std::uint16_t kVersion = 1;
constexpr std::size_t kHeaderSize = 12;

[[nodiscard]] Result<void> readFully(BinaryReader& reader, std::span<std::byte> destination) {
    std::size_t offset = 0;
    while (offset < destination.size()) {
        auto got = reader.read(destination.subspan(offset));
        if (!got.has_value()) {
            return std::unexpected(got.error());
        }
        if (*got == 0) {
            return std::unexpected(Error{ErrorCode::InvalidArgument, "memory: truncated document"});
        }
        offset += *got;
    }
    return {};
}

[[nodiscard]] Result<std::uint16_t> readU16le(const std::array<std::byte, kHeaderSize>& h,
                                              std::size_t at) {
    return static_cast<std::uint16_t>((std::to_integer<unsigned int>(h[at + 0]) << 0) |
                                      (std::to_integer<unsigned int>(h[at + 1]) << 8));
}

[[nodiscard]] Result<std::uint32_t> readU32le(const std::array<std::byte, kHeaderSize>& h,
                                              std::size_t at) {
    return static_cast<std::uint32_t>((std::to_integer<unsigned int>(h[at + 0]) << 0) |
                                      (std::to_integer<unsigned int>(h[at + 1]) << 8) |
                                      (std::to_integer<unsigned int>(h[at + 2]) << 16) |
                                      (std::to_integer<unsigned int>(h[at + 3]) << 24));
}

/// @brief Writes a little-endian u16 to \p writer at the cursor.
[[nodiscard]] Result<void> writeU16le(BinaryWriter& writer, std::uint16_t value) {
    const std::byte raw[2]{static_cast<std::byte>((value >> 0) & 0xFFu),
                           static_cast<std::byte>((value >> 8) & 0xFFu)};
    return writer.write(std::span<const std::byte>{raw, 2})
        .and_then([](std::size_t) -> Result<void> { return {}; });
}

/// @brief Writes a little-endian u32 to \p writer at the cursor.
[[nodiscard]] Result<void> writeU32le(BinaryWriter& writer, std::uint32_t value) {
    const std::byte raw[4]{static_cast<std::byte>((value >> 0) & 0xFFu),
                           static_cast<std::byte>((value >> 8) & 0xFFu),
                           static_cast<std::byte>((value >> 16) & 0xFFu),
                           static_cast<std::byte>((value >> 24) & 0xFFu)};
    return writer.write(std::span<const std::byte>{raw, 4})
        .and_then([](std::size_t) -> Result<void> { return {}; });
}

/// @brief Parses the 12-byte PMEM header from \p reader (cursor moved to 0), validates it, and
///        reports the length of the model header that follows.
struct DocumentHeader {
    std::uint32_t modelLen = 0;
};

[[nodiscard]] Result<DocumentHeader> readDocumentHeader(BinaryReader& reader) {
    auto seekResult = reader.seek(0);
    if (!seekResult.has_value()) {
        return std::unexpected(seekResult.error());
    }
    auto size = reader.size();
    if (!size.has_value()) {
        return std::unexpected(size.error());
    }
    std::array<std::byte, kHeaderSize> header{};
    auto readResult = readFully(reader, std::span<std::byte>{header});
    if (!readResult.has_value()) {
        return std::unexpected(readResult.error());
    }
    for (std::size_t i = 0; i < kMagic.size(); ++i) {
        if (header[i] != kMagic[i]) {
            return std::unexpected(Error{ErrorCode::InvalidArgument, "memory: bad magic"});
        }
    }
    auto version = readU16le(header, 4);
    if (version != kVersion) {
        return std::unexpected(Error{ErrorCode::InvalidArgument, "memory: unsupported version"});
    }
    auto modelLen = readU32le(header, 8);
    if (static_cast<std::uint64_t>(*modelLen) + kHeaderSize > *size) {
        return std::unexpected(
            Error{ErrorCode::InvalidArgument, "memory: model header exceeds file size"});
    }
    return DocumentHeader{*modelLen};
}

/// @brief Reads and decodes the whole document (header + model tree) from \p reader.
struct ParsedDocument {
    io::StorageModel root;
    std::uint64_t docPrefix = 0; // absolute offset where pixel regions begin
};

[[nodiscard]] Result<ParsedDocument> readDocument(BinaryReader& reader) {
    auto header = readDocumentHeader(reader);
    if (!header.has_value()) {
        return std::unexpected(header.error());
    }
    std::vector<std::byte> model(static_cast<std::size_t>(header->modelLen));
    auto readResult = readFully(reader, std::span<std::byte>{model});
    if (!readResult.has_value()) {
        return std::unexpected(readResult.error());
    }
    std::size_t cursor = 0;
    std::uint64_t nodeBudget = 0;
    auto root = memory::memoryDecode(model, cursor, nodeBudget);
    if (!root.has_value()) {
        return std::unexpected(root.error());
    }
    return ParsedDocument{std::move(*root),
                          static_cast<std::uint64_t>(kHeaderSize) + header->modelLen};
}

/// @brief Writes the 12-byte PMEM header then \p modelBytes (the encoded model tree) at cursor 0.
[[nodiscard]] Result<void> writeHeaderAndModel(BinaryWriter& writer,
                                               std::span<const std::byte> modelBytes) {
    auto seekResult = writer.seek(0);
    if (!seekResult.has_value()) {
        return std::unexpected(seekResult.error());
    }
    auto magicResult = writer.write(kMagic);
    if (!magicResult.has_value()) {
        return std::unexpected(magicResult.error());
    }
    auto versionResult = writeU16le(writer, kVersion);
    if (!versionResult.has_value()) {
        return std::unexpected(versionResult.error());
    }
    auto flagsResult = writeU16le(writer, 0);
    if (!flagsResult.has_value()) {
        return std::unexpected(flagsResult.error());
    }
    auto lenResult = writeU32le(writer, static_cast<std::uint32_t>(modelBytes.size()));
    if (!lenResult.has_value()) {
        return std::unexpected(lenResult.error());
    }
    return writer.write(modelBytes).and_then([](std::size_t) -> Result<void> { return {}; });
}

/// @brief Derives per-image storage info for a set of flat image models.
[[nodiscard]] Result<std::vector<memory::MemoryImageInfo>>
imageInfos(std::span<const io::StorageModel> models) {
    std::vector<memory::MemoryImageInfo> infos;
    infos.reserve(models.size());
    for (const auto& model : models) {
        auto info = memory::imageInfoFromModel(model);
        if (!info.has_value()) {
            return std::unexpected(info.error());
        }
        infos.push_back(std::move(*info));
    }
    return infos;
}

} // namespace

std::string_view MemoryBackend::name() const noexcept {
    return "memory";
}

BackendCapabilities MemoryBackend::capabilities() const noexcept {
    return {.supportsTiling = true,
            .supportsStreaming = false,
            .supportsRandomAccess = true,
            .supportsCloudStreaming = false};
}

Result<std::unique_ptr<ImageSource>> MemoryBackend::openImageSource(BinaryReader& reader) const {
    return openImageSourceAt(reader, 0);
}

Result<std::unique_ptr<ImageSink>> MemoryBackend::openImageSink(BinaryWriter& writer,
                                                                const StorageModel& model) const {
    auto children = model.children();
    if (children.empty()) {
        return std::unexpected(Error{ErrorCode::InvalidArgument,
                                     "memory: openImageSink requires a model with images"});
    }
    auto docPrefix = static_cast<std::uint64_t>(kHeaderSize) + memory::memoryCodecByteSize(model);
    auto info = memory::imageInfoFromModel(children.front());
    if (!info.has_value()) {
        return std::unexpected(info.error());
    }
    auto seekResult = writer.seek(docPrefix);
    if (!seekResult.has_value()) {
        return std::unexpected(seekResult.error());
    }
    return std::make_unique<memory::MemoryImageSink>(writer, std::move(*info), docPrefix);
}

Result<StorageModel> MemoryBackend::deserializeModel(BinaryReader& reader) const {
    auto doc = readDocument(reader);
    if (!doc.has_value()) {
        return std::unexpected(doc.error());
    }
    return std::move(doc->root);
}

Result<void> MemoryBackend::serializeModel(const StorageModel& model, BinaryWriter& writer) const {
    std::vector<std::byte> bytes;
    std::uint64_t nodeBudget = 0;
    auto encodeResult = memory::memoryEncode(bytes, model, nodeBudget);
    if (!encodeResult.has_value()) {
        return std::unexpected(encodeResult.error());
    }
    return writeHeaderAndModel(writer, bytes);
}

Result<void> MemoryBackend::serializeModelList(std::span<const StorageModel> models,
                                               BinaryWriter& writer) const {
    std::vector<std::byte> bytes;
    std::uint64_t nodeBudget = 0;
    auto encodeResult = memory::memoryEncodeDocument(bytes, models, nodeBudget);
    if (!encodeResult.has_value()) {
        return std::unexpected(encodeResult.error());
    }
    return writeHeaderAndModel(writer, bytes);
}

Result<std::unique_ptr<ImageSink>> MemoryBackend::openImageSinkAt(
    BinaryWriter& writer, std::span<const StorageModel> models, std::size_t imageIndex) const {
    if (imageIndex >= models.size()) {
        return std::unexpected(Error{ErrorCode::OutOfRange,
                                     "MemoryBackend::openImageSinkAt: image index out of range"});
    }
    auto infos = imageInfos(models);
    if (!infos.has_value()) {
        return std::unexpected(infos.error());
    }
    const std::uint64_t docPrefix =
        static_cast<std::uint64_t>(kHeaderSize) + memory::memoryDocumentByteSize(models);
    const std::uint64_t pixelStart = memory::imagePixelOffset(*infos, imageIndex, docPrefix);
    auto seekResult = writer.seek(pixelStart);
    if (!seekResult.has_value()) {
        return std::unexpected(seekResult.error());
    }
    return std::make_unique<memory::MemoryImageSink>(
        writer, std::move((*infos)[imageIndex]), pixelStart);
}

Result<std::unique_ptr<ImageSource>>
MemoryBackend::openImageSourceAt(BinaryReader& reader, std::size_t imageIndex) const {
    auto doc = readDocument(reader);
    if (!doc.has_value()) {
        return std::unexpected(doc.error());
    }
    auto children = doc->root.children();
    if (imageIndex >= children.size()) {
        return std::unexpected(Error{ErrorCode::OutOfRange,
                                     "MemoryBackend::openImageSourceAt: image index out of range"});
    }
    auto infos = imageInfos(children);
    if (!infos.has_value()) {
        return std::unexpected(infos.error());
    }
    const std::uint64_t pixelStart = memory::imagePixelOffset(*infos, imageIndex, doc->docPrefix);
    return std::make_unique<memory::MemoryImageSource>(
        reader, std::move((*infos)[imageIndex]), pixelStart);
}

namespace {
const bool registered = [] {
    return BackendFactory::instance()
        .registerBackend("memory", [] { return std::make_unique<MemoryBackend>(); })
        .has_value();
}();
} // namespace

} // namespace ptiff::io::backend

File: crates/ptiff-core/src/io/backend/memory_backend.rs
//! In-memory ("PMEM") [`StorageBackend`].
//!
//! Rust-first implementation (no byte-for-byte C++ PMEM binary format): the
//! format-neutral [`StorageModel`] is encoded to/from JSON via `serde_json`
//! (§4.4/§4.6 of PTIFF-1.0-RUST-CORE-PLAN.md — "Eigenes JSON-Parsing → serde").
//! A JSON-encoded model occupies the "document header" region before the
//! contiguous per-image pixel regions, which [`MemoryImageSource`] /
//! [`MemoryImageSink`] address at linear tile offsets.
//!
//! This module is only compiled when the `memory-backend` feature is enabled.

use crate::io::{
    BackendCapabilities, BinaryReader, BinaryWriter, ImageSink, ImageSource, StorageBackend,
};
use crate::{Error, Result, StorageModel};

use super::memory_image_sink::MemoryImageSink;
use super::memory_image_source::MemoryImageSource;
use super::memory_layout::{image_info_from_model, image_pixel_offset, MemoryImageInfo};

/// The in-memory storage backend.
///
/// Persistent behavior is provided via the [`BinaryReader`]/[`BinaryWriter`]
/// transports (typically [`crate::io::MemoryBinaryReader`] /
/// [`crate::io::MemoryBinaryWriter`]) with the model encoded as JSON.
///
/// Capabilities: the format supports tiled layouts and arbitrary seek (random
/// access); it is not cloud-safe — it streams over a whole in-memory document.
pub struct MemoryBackend;

impl MemoryBackend {
    /// Encodes `model` as the JSON bytes that precede its pixel region.
    fn model_json_bytes(model: &StorageModel) -> Result<Vec<u8>> {
        serde_json::to_vec(model)
            .map_err(|e| Error::unknown(format!("MemoryBackend: encode failed: {e}")))
    }

    /// Writes `bytes` in full to `writer`, failing on a short write.
    fn write_all(writer: &mut dyn BinaryWriter, bytes: &[u8]) -> Result<()> {
        let mut written = 0usize;
        while written < bytes.len() {
            let n = writer.write(&bytes[written..])?;
            if n == 0 {
                return Err(Error::unknown(
                    "MemoryBackend: transport refused to accept bytes",
                ));
            }
            written += n;
        }
        writer.flush()?;
        Ok(())
    }

    /// Reads the entire transport into a buffer until read returns 0 bytes.
    fn read_whole(reader: &mut dyn BinaryReader) -> Result<Vec<u8>> {
        let mut buf = Vec::new();
        let mut chunk = [0u8; 4096];
        loop {
            let n = reader.read(&mut chunk)?;
            buf.extend_from_slice(&chunk[..n]);
            if n == 0 {
                break;
            }
        }
        Ok(buf)
    }

    /// Derives per-image info for a list of flat image models.
    fn image_infos(models: &[StorageModel]) -> Result<Vec<MemoryImageInfo>> {
        models.iter().map(image_info_from_model).collect()
    }
}

impl StorageBackend for MemoryBackend {
    fn name(&self) -> &'static str {
        "memory"
    }

    fn capabilities(&self) -> BackendCapabilities {
        BackendCapabilities {
            supports_tiling: true,
            supports_streaming: true,
            supports_random_access: true,
            supports_cloud_streaming: false,
        }
    }

    fn open_image_source<'a>(
        &self,
        reader: &'a mut dyn BinaryReader,
    ) -> Result<Box<dyn ImageSource + 'a>> {
        // A single-image document has a JSON-object header (from `serialize_model`)
        // whose children are the image models. Decode the leading object and
        // open image 0.
        let root = self.deserialize_model(reader)?;
        let children = root.children();
        if children.is_empty() {
            return Err(Error::invalid_argument(
                "memory: openImageSource requires a model with images",
            ));
        }
        let infos = Self::image_infos(children)?;
        let doc_prefix = Self::model_json_bytes(&root)?.len() as u64;
        let pixel_start = image_pixel_offset(&infos, 0, doc_prefix);
        Ok(Box::new(MemoryImageSource::new(
            reader,
            infos.into_iter().next().unwrap(),
            pixel_start,
        )))
    }

    fn open_image_sink<'a>(
        &self,
        writer: &'a mut dyn BinaryWriter,
        model: &StorageModel,
    ) -> Result<Box<dyn ImageSink + 'a>> {
        let children = model.children();
        if children.is_empty() {
            return Err(Error::invalid_argument(
                "memory: openImageSink requires a model with images",
            ));
        }
        // The pixel region begins right after this model's JSON header bytes.
        let doc_prefix = Self::model_json_bytes(model)?.len() as u64;
        let info = image_info_from_model(&children[0])?;
        writer.seek(doc_prefix)?;
        Ok(Box::new(MemoryImageSink::new(writer, info, doc_prefix)))
    }

    fn deserialize_model(&self, reader: &mut dyn BinaryReader) -> Result<StorageModel> {
        // A memory document is "JSON model header" followed by the contiguous
        // per-image pixel regions. Read the whole transport then decode only
        // the leading JSON object, ignoring any trailing pixel bytes.
        let buf = Self::read_whole(reader)?;
        let mut stream = serde_json::Deserializer::from_slice(&buf).into_iter::<StorageModel>();
        stream
            .next()
            .ok_or_else(|| Error::invalid_argument("MemoryBackend: bad model: empty document"))?
            .map_err(|e| Error::invalid_argument(format!("MemoryBackend: bad model: {e}")))
    }

    fn serialize_model(&self, model: &StorageModel, writer: &mut dyn BinaryWriter) -> Result<()> {
        let bytes = Self::model_json_bytes(model)?;
        Self::write_all(writer, &bytes)
    }

    fn serialize_model_list(
        &self,
        models: &[StorageModel],
        writer: &mut dyn BinaryWriter,
    ) -> Result<()> {
        let bytes = serde_json::to_vec(models)
            .map_err(|e| Error::unknown(format!("MemoryBackend: encode failed: {e}")))?;
        Self::write_all(writer, &bytes)
    }

    fn open_image_sink_at<'a>(
        &self,
        writer: &'a mut dyn BinaryWriter,
        models: &[StorageModel],
        image_index: usize,
    ) -> Result<Box<dyn ImageSink + 'a>> {
        if image_index >= models.len() {
            return Err(Error::out_of_range(
                "MemoryBackend::openImageSinkAt: image index out of range",
            ));
        }
        let infos = Self::image_infos(models)?;
        // The multi-image document's pixel regions begin after the JSON array
        // that encodes every model.
        let header_bytes = serde_json::to_vec(models)
            .map_err(|e| Error::unknown(format!("MemoryBackend: encode failed: {e}")))?;
        let doc_prefix = header_bytes.len() as u64;
        let pixel_start = image_pixel_offset(&infos, image_index, doc_prefix);
        writer.seek(pixel_start)?;
        Ok(Box::new(MemoryImageSink::new(
            writer,
            infos.into_iter().nth(image_index).unwrap(),
            pixel_start,
        )))
    }

    fn open_image_source_at<'a>(
        &self,
        reader: &'a mut dyn BinaryReader,
        image_index: usize,
    ) -> Result<Box<dyn ImageSource + 'a>> {
        // A multi-image document has a JSON-array header of models followed by
        // the contiguous per-image pixel regions. Decode only the leading array
        // value, ignoring any trailing pixel bytes.
        let buf = Self::read_whole(reader)?;
        let mut stream =
            serde_json::Deserializer::from_slice(&buf).into_iter::<Vec<StorageModel>>();
        let models = stream
            .next()
            .ok_or_else(|| Error::invalid_argument("MemoryBackend: bad model: empty document"))?
            .map_err(|e| Error::invalid_argument(format!("MemoryBackend: bad model: {e}")))?;
        if image_index >= models.len() {
            return Err(Error::out_of_range(
                "MemoryBackend::openImageSourceAt: image index out of range",
            ));
        }
        let infos = Self::image_infos(&models)?;
        let doc_prefix = serde_json::to_vec(&models)
            .map_err(|e| Error::unknown(format!("MemoryBackend: encode failed: {e}")))?
            .len() as u64;
        let pixel_start = image_pixel_offset(&infos, image_index, doc_prefix);
        Ok(Box::new(MemoryImageSource::new(
            reader,
            infos.into_iter().nth(image_index).unwrap(),
            pixel_start,
        )))
    }
}

impl Default for MemoryBackend {
    fn default() -> Self {
        Self
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::id::TileId;
    use crate::io::MemoryBinaryReader;
    use crate::io::MemoryBinaryWriter;
    use crate::tile::{Tile, TileExtent, TileIndex, TileLayout, TileRegion};
    use crate::ErrorCode;

    fn sample_model() -> StorageModel {
        let mut root = StorageModel::new();
        root.set_field("producer", "ptiff-core");
        let mut image = StorageModel::new();
        image.set_field("imageWidth", "64");
        image.set_field("imageHeight", "32");
        image.set_field("tileWidth", "16");
        image.set_field("tileHeight", "16");
        image.set_field("samplesPerPixel", "1");
        image.set_field("pixelType", "UInt8");
        image.set_field("compression", "None");
        root.add_child(image);
        root
    }

    fn tiled_sink_payload() -> Vec<u8> {
        vec![0u8; 256]
    }

    #[test]
    fn name_and_capabilities() {
        let backend = MemoryBackend;
        assert_eq!(backend.name(), "memory");
        let caps = backend.capabilities();
        assert!(caps.supports_streaming);
        assert!(caps.supports_random_access);
        assert!(caps.supports_tiling);
        assert!(!caps.supports_cloud_streaming);
    }

    #[test]
    fn serialize_then_deserialize_round_trips_model() {
        let backend = MemoryBackend;
        let model = sample_model();

        let mut writer = MemoryBinaryWriter::new();
        backend
            .serialize_model(&model, &mut writer)
            .expect("serialize");
        assert!(!writer.buffer().is_empty());

        let bytes = writer.take_buffer();
        let mut reader = MemoryBinaryReader::from_vec(bytes);
        let back = backend.deserialize_model(&mut reader).expect("deserialize");

        // Fields round-trip (BTreeMap keeps canonical field order in JSON).
        assert_eq!(back.field("producer").unwrap(), "ptiff-core");
        assert_eq!(back.child_count(), 1);
        assert_eq!(back.children()[0].field("imageWidth").unwrap(), "64");
    }

    #[test]
    fn deserialize_garbage_is_invalid_argument() {
        let backend = MemoryBackend;
        let mut writer = MemoryBinaryWriter::new();
        writer.write(b"not-json").unwrap();
        let bytes = writer.take_buffer();
        let mut reader = MemoryBinaryReader::from_vec(bytes);
        let err = backend
            .deserialize_model(&mut reader)
            .expect_err("must fail");
        assert_eq!(err.code(), ErrorCode::InvalidArgument);
    }

    #[test]
    fn open_image_sink_requires_a_model_with_images() {
        let backend = MemoryBackend;
        let mut writer = MemoryBinaryWriter::new();
        let empty = StorageModel::new();
        let err = match backend.open_image_sink(&mut writer, &empty) {
            Ok(_) => panic!("expected error"),
            Err(e) => e,
        };
        assert_eq!(err.code(), ErrorCode::InvalidArgument);
        assert_eq!(
            err.message(),
            "memory: openImageSink requires a model with images"
        );
    }

    #[test]
    fn pixel_sink_writes_and_source_reads_back() {
        let backend = MemoryBackend;
        let model = sample_model();

        // 1) serialize the model header so the document header exists.
        let mut writer = MemoryBinaryWriter::new();
        backend.serialize_model(&model, &mut writer).unwrap();
        // Append the (zero-filled) pixel region right after the JSON header so
        // the sink can seek to an in-range tile offset. The cursor is already
        // past the header, so writing the pixel-region bytes appends them.
        let info = {
            use crate::io::backend::memory_layout::image_info_from_model;
            image_info_from_model(&model.children()[0]).unwrap()
        };
        {
            use crate::io::BinaryWriter as _;
            let zeros = vec![0u8; info.image_pixel_bytes as usize];
            writer.write(&zeros).unwrap();
        }

        // 2) open a sink over the same (already header-holding) writer.
        let mut sink = backend.open_image_sink(&mut writer, &model).unwrap();
        let mut payload = tiled_sink_payload();
        payload[0] = 0x42;
        payload[255] = 0x24;
        let tile = Tile::new(
            TileId::new(0),
            TileIndex::new(1, 0, 0),
            TileRegion::new(16, 0, TileExtent::new(16, 16)),
            &payload,
        );
        sink.write_tile(&tile).unwrap();
        assert_eq!(sink.layout().image_width, 64);
        drop(sink);

        // 3) read back via source. Tile (1,0) is the 2nd tile, at the JSON
        // header length + 256 bytes; the round-trip values prove the offset.
        let bytes = writer.take_buffer();
        let mut reader = MemoryBinaryReader::from_vec(bytes);
        let mut source = backend.open_image_source(&mut reader).unwrap();
        assert_eq!(
            source.layout(),
            &TileLayout::new(TileExtent::new(16, 16), 64, 32, 1)
        );
        let back = source.read_tile(TileIndex::new(1, 0, 0)).unwrap();
        assert_eq!(back.data()[0], 0x42);
        assert_eq!(back.data()[255], 0x24);
        assert_eq!(back.index(), TileIndex::new(1, 0, 0));
    }

    #[test]
    fn open_image_source_at_index_out_of_range() {
        let backend = MemoryBackend;
        let model = sample_model();
        // A multi-image document is written as a JSON array of models.
        let list = vec![model];
        let header = serde_json::to_vec(&list).unwrap();
        let bytes = header; // header alone is a valid array document (no pixels)
        let mut reader = MemoryBinaryReader::from_vec(bytes);
        let err = match backend.open_image_source_at(&mut reader, 99) {
            Ok(_) => panic!("expected error"),
            Err(e) => e,
        };
        assert_eq!(err.code(), ErrorCode::OutOfRange);
    }

    #[test]
    fn multi_image_sink_at_and_source_at_use_strided_offsets() {
        let backend = MemoryBackend;
        // Two images; the second's pixel region follows the first's.
        let mut a = StorageModel::new();
        a.set_field("imageWidth", "64");
        a.set_field("imageHeight", "32");
        a.set_field("tileWidth", "16");
        a.set_field("tileHeight", "16");
        a.set_field("samplesPerPixel", "1");
        a.set_field("pixelType", "UInt8");
        a.set_field("compression", "None");
        let mut b = StorageModel::new();
        b.set_field("imageWidth", "32");
        b.set_field("imageHeight", "16");
        b.set_field("tileWidth", "16");
        b.set_field("tileHeight", "16");
        b.set_field("samplesPerPixel", "1");
        b.set_field("pixelType", "UInt8");
        b.set_field("compression", "None");

        let models = vec![a, b];
        let infos = {
            use crate::io::backend::memory_layout::image_info_from_model;
            models
                .iter()
                .map(image_info_from_model)
                .collect::<Result<Vec<_>>>()
                .unwrap()
        };
        // Multi-image header is the JSON array of all models.
        let header = serde_json::to_vec(&models).unwrap();
        let doc_prefix = header.len() as u64;
        let pixel_start_a = doc_prefix;
        let pixel_start_b = doc_prefix + infos[0].image_pixel_bytes;

        // Pre-size the buffer: write the real JSON-array header, then
        // zero-fill the two pixel regions after it so the sink can seek in-range.
        let total = doc_prefix as usize
            + infos[0].image_pixel_bytes as usize
            + infos[1].image_pixel_bytes as usize;
        let mut writer = MemoryBinaryWriter::new();
        {
            use crate::io::BinaryWriter as _;
            writer.write(&header).unwrap();
            let zeros = vec![0u8; total - header.len()];
            writer.write(&zeros).unwrap();
        }

        // Write the *second* image's tile (0,0) via sink_at.
        let mut sink_b = backend.open_image_sink_at(&mut writer, &models, 1).unwrap();
        let mut payload_b = vec![0u8; 256];
        payload_b[0] = 0xEE;
        sink_b
            .write_tile(&Tile::new(
                TileId::new(0),
                TileIndex::new(0, 0, 0),
                TileRegion::new(0, 0, TileExtent::new(16, 16)),
                &payload_b,
            ))
            .unwrap();
        drop(sink_b);

        let bytes = writer.take_buffer();
        // Read back image index 1 from the source_at path.
        let mut reader = MemoryBinaryReader::from_vec(bytes.clone());
        let mut source_b = backend.open_image_source_at(&mut reader, 1).unwrap();
        let back_b = source_b.read_tile(TileIndex::new(0, 0, 0)).unwrap();
        assert_eq!(back_b.data()[0], 0xEE);
        // Verify indexing into the raw buffer: image 1's first tile sits at
        // pixel_start_b.
        assert_eq!(
            bytes[pixel_start_b as usize], 0xEE,
            "image 1 tile must land at its strided pixel offset"
        );
        // Verify image 0's region is still all zeros (untouched).
        assert_eq!(bytes[pixel_start_a as usize], 0);

        // open_image_sink_at out-of-range guard.
        let mut w2 = MemoryBinaryWriter::new();
        let err = match backend.open_image_sink_at(&mut w2, &models, 99) {
            Ok(_) => panic!("expected error"),
            Err(e) => e,
        };
        assert_eq!(err.code(), ErrorCode::OutOfRange);
    }
}
